storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,22 +3571,429 @@ dberr_t fsp_sys_tablespace_validate()
+  bool insert_inode(uint32_t page, uint16_t offset)
+  {
+    return page < fil_system.sys_space->free_limit &&
+      offset >= FIL_PAGE_DATA && offset < srv_page_size - FIL_PAGE_DATA_END &&
+      emplace(uint64_t{page} << 32 | offset).second;
+  }
+
+  /** Register an inode
+  @param inode index node information
+  @retval true in case of successful registeration
+  @retval false in case of invalid entry or already inserted inode */
+  __attribute__((warn_unused_result))
+  bool insert_seg(const byte *inode)
+  {
+    return insert_inode(mach_read_from_4(inode + 4),
+                        mach_read_from_2(inode + 8));
+  }
+
+  __attribute__((warn_unused_result))
+  bool find(uint32_t page, uint16_t offset) const
+  {
+    return std::unordered_set<uint64_t>::find(uint64_t{page} << 32 |
+                                              offset) != end();
+  }
+
+  /** Get the unused inode segment header from the list of index
+  node pages.
+  @param boffset offset for the FSP_SEG_INODES_FULL
+                 or FSP_SEG_INODES_FREE list in fsp header page
+  @param unused  store the unused information
+  @return error code */
+  dberr_t get_unused(uint16_t boffset, inode_info *unused) const
+  {
+    dberr_t err= DB_SUCCESS;
+    buf_block_t *block= buf_pool.page_fix(page_id_t{0, 0}, &err,
+                                          buf_pool_t::FIX_WAIT_READ);
+    if (!block)
+      return err;
+    buf_block_t *header= block;
+    const uint32_t len= flst_get_len(block->page.frame + boffset);
+    fil_addr_t addr= flst_get_first(block->page.frame + boffset);
+    ulint n_inode_per_page=
+      FSP_SEG_INODES_PER_PAGE(fil_system.sys_space->physical_size());
+    for (uint32_t i= len; i--; )
+    {
+      if (addr.boffset < FIL_PAGE_DATA ||
+          addr.boffset >= block->physical_size() - FIL_PAGE_DATA_END)
+      {
+        err= DB_CORRUPTION;
+        break;
+      }
+
+      block= buf_pool.page_fix(page_id_t{0, addr.page}, &err,
+                               buf_pool_t::FIX_WAIT_READ);
+      if (!block)
+        break;
+
+      fil_addr_t next_addr= flst_get_next_addr(block->page.frame +
+                                               addr.boffset);
+      for (uint32_t i= 0; i < n_inode_per_page; i++)
+      {
+        const fseg_inode_t *inode=
+          fsp_seg_inode_page_get_nth_inode(block->page.frame, i);
+        ulint seg_id= mach_read_from_8(FSEG_ID + inode);
+        /* Consider TRX_SYS_FSEG_HEADER as used segment.
+        While reinitializing the undo tablespace, InnoDB
+        fail to reset the value of TRX_SYS_FSEG_HEADER
+        in TRX_SYS page. so InnoDB shouldn't consider
+        this segment as unused one */
+        if (seg_id == 0 || seg_id == 2)
+          continue;
+	uint16_t offset= uint16_t(inode - block->page.frame);
+        if (!find(addr.page, offset) &&
+            !unused->insert_inode(addr.page, offset))
+        {
+          err= DB_CORRUPTION;
+          break;
+        }
```

Comment:
Please distinguish these as `DB_RECORD_NOT_FOUND` and `DB_DUPLICATE_KEY` so that we can output more accurate messages.

Reply:
Sorry, I misread the condition as `||` instead of `&&`. We can only return one error here, `DB_DUPLICATE_KEY`.

Reply:
There is no need for `DB_RECORD_NOT_FOUND`